tests/ui/cypress/specs/multi_user_openid_connect.js

```diff
@@ -0,0 +1,289 @@
+
+        cy.contains('Sign In').click()
+
+        // We should end up back in the Krill UI
+        cy.url().should('include', Cypress.config('baseUrl'))
+
+        if (ts.o) {
+          cy.contains('Sign In').should('not.exist')
+          cy.get('#userinfo').click()
+          cy.get('#userinfo_table').contains(ts.u)
+          cy.get('#userinfo_table').contains('role')
+        } else if (ts.d == 'empty') {
+          cy.contains('The supplied login credentials were incorrect')
+          cy.contains('return to the login page')
+        } else if (ts.d == 'badidtoken') {
+          cy.contains('OpenID Connect: Code exchange failed: Failed to parse server response')
+          cy.contains('return to the login page')
+        } else if (ts.d == 'badrole') {
+          cy.contains(
+            'Your user does not have sufficient rights to perform this action. Please contact your administrator.'
+          )
+          cy.contains('return to the login page')
+        }
+      }
+    )
+  })
+
+  it('Can logout', () => {
+    // login
+    cy.visit('/')
+    cy.url().should('not.include', Cypress.config('baseUrl'))
+    cy.contains('Mock OpenID Connect login form')
+    cy.get('input[name="username"]').clear().type(admin.u)
+    cy.contains('Sign In').click()
+
+    // verify that we are shown to be logged in to the Krill UI
+    cy.contains('Sign In').should('not.exist')
+    cy.url().should('include', Cypress.config('baseUrl'))
+    cy.get('#userinfo').click()
+    cy.get('#userinfo_table').contains(admin.u)
+
+    // logout
+    cy.intercept('GET', /^http:\/\/localhost:1818\/logout.+/).as('oidcLogout')
+    cy.get('.logout').click()
+    cy.wait('@oidcLogout').its('response.statusCode').should('eq', 302)
+
+    // verify that we are shown the OpenID Connect provider login page
+    cy.url().should('not.include', Cypress.config('baseUrl'))
+    cy.contains('Mock OpenID Connect login form')
+    cy.get('input[name="username"]')
+  })
+
+  it('Login with short-lived non-refreshable token and try to refresh page', () => {
+    cy.intercept('GET', '/api/v1/authorized').as('isAuthorized')
+    cy.visit('/')
+
+    cy.wait('@isAuthorized').its('response.statusCode').should('eq', 403)
+    cy.url().should('not.include', Cypress.config('baseUrl'))
+    cy.contains('Mock OpenID Connect login form')
+    cy.get('input[name="username"]').clear().type(shorttoken.u)
+
+    cy.intercept('GET', '/index.html').as('postLoginIndexFetch')
+    cy.intercept('GET', '/api/v1/authorized').as('isAuthorized')
+    cy.contains('Sign In').click()
+
+    cy.wait('@postLoginIndexFetch').its('response.statusCode').should('eq', 200)
+    cy.wait('@isAuthorized').its('response.statusCode').should('eq', 200)
+    cy.url().should('include', Cypress.config('baseUrl'))
+    cy.contains('Sign In').should('not.exist')
+    cy.get('#userinfo').click()
+    cy.get('#userinfo_table').contains(shorttoken.u)
+    cy.contains(shorttoken.u)
```

Comment:
In `multi_user_config_file_with_ta.js` this is done differently:

```
// create a CA
cy.contains('CA Handle')
cy.get('form input[type="text"]').type(ts.ca)
cy.contains('Create CA').click()
cy.contains('OK').click()
```

The intercept you have added here would be good to add in the `with_ta` JS script as a future improvement (let's not mix changes to that test in with this PR, I just want to say I like your use of the intercept here).

However, I find the locators used in the `with_ta` test JS script easier to understand (click "Create CA" instead of click "main button" for example) so perhaps it would be good to use the locator style from the `with_ta` JS script here.

Reply:
Fair enough, I'll change it to do that. I wish though that Lagosta would have meaningful id attributes on the html elements involved. Selecting based on textContent is still a bit brittle.

Reply:
We can add them, I have done so in some cases while working on the multi-user work. A related approach that I stumbled upon yesterday is to add `cy-name` attributes.